ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/PrintSubscriber.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2022 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.api.server;
+
+import static io.confluent.ksql.rest.Errors.ERROR_CODE_SERVER_ERROR;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.confluent.ksql.reactive.BaseSubscriber;
+import io.confluent.ksql.rest.entity.KsqlErrorMessage;
+import io.vertx.core.Context;
+import io.vertx.core.http.HttpServerResponse;
+import java.util.Objects;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is a reactive streams subscriber which receives a stream of results from a publisher which
+ * is implemented by the back-end. The results are then written to the HTTP2 response.
+ */
+public class PrintSubscriber extends BaseSubscriber<String> {
+
+  private static final Logger log = LoggerFactory.getLogger(PrintSubscriber.class);
+  private static final int REQUEST_BATCH_SIZE = 200;
+
+  private final HttpServerResponse response;
+  private final DelimitedPrintResponseWriter printResponseWriter;
+  private int tokens;
+
+  @SuppressFBWarnings(value = "EI_EXPOSE_REP2")
+  public PrintSubscriber(final Context context, final HttpServerResponse response,
+      final DelimitedPrintResponseWriter printResponseWriter) {
+    super(context);
+    this.response = Objects.requireNonNull(response);
+    this.printResponseWriter = Objects.requireNonNull(printResponseWriter);
+  }
+
+  @Override
+  protected void afterSubscribe(final Subscription subscription) {
+    checkMakeRequest();
+  }
+
+  @Override
+  public void handleValue(final String row) {
+    printResponseWriter.writeRow(row);
+    tokens--;
+    if (response.writeQueueFull()) {
+      response.drainHandler(v -> checkMakeRequest());
+    } else {
+      checkMakeRequest();
+    }
+  }
+
+  private void checkMakeRequest() {
+    if (tokens == 0) {
+      tokens = REQUEST_BATCH_SIZE;
+      makeRequest(REQUEST_BATCH_SIZE);
+    }
+  }
+
+  @Override
+  public void handleError(final Throwable t) {
```

Comment:
Does your publisher ever actually send errors? If not then that means the only methods of PrintResponseWriter which are actually used are writeRow() and end(). If that's the case, let's simply this code by removing the interface entirely and inlining the implementation, unless there are future plans to introduce other formats.

Reply:
You're right, this is unnecessary, and I like the idea of simplifying things! I'm removing the PrintResponseWriter altogether and moving the response.write logic to the PrintSubscriber.

And we do send a different error message when the topic doesn't exist, for example, but we also don't need handleError for that.